internal/lines.go

```diff
@@ -0,0 +1,100 @@
+package internal
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type LineHandler struct {
+	Reporter      Reporter
+	BatchSize     int
+	MaxBufferSize int
+	FlushTicker   *time.Ticker
+	Format        string
+	failures      int64
+	mtx           sync.Mutex
+	lines         []string
+}
+
+func (lh *LineHandler) Start() {
+	go func() {
+		for range lh.FlushTicker.C {
+			lh.Flush()
+		}
+	}()
+}
+
+func (lh *LineHandler) HandleLine(line string) error {
+	lh.mtx.Lock()
+	defer lh.mtx.Unlock()
+
+	if len(lh.lines) >= lh.MaxBufferSize {
+		atomic.AddInt64(&lh.failures, 1)
+		return fmt.Errorf("buffer full, dropping line: %s", line)
```

Comment:
Wouldn't it be better to try to flush the buffer instead of dropping the line? The whole design of this class would be a lot easier if you used a channel as the underlying buffer. Then ```HandleLine()``` could just send something on the channel and you could have two flavors of ```HandleLine```: One that blocks if the buffer is full and one that drops. Also, this would eliminate the need for the mutex.